File: cbus_bnb.py
"""
Branch-and-Bound solver for CBUS (Pickup and Delivery Problem)
Python port of cbus_bnb.cpp

Các ràng buộc:
  * Mỗi điểm trong {1..2n} ghé đúng một lần, xuất phát và quay về 0
  * Pickup i ghé trước Delivery i+n
  * Sức chứa [0, k] tại mọi bước
  * Tối thiểu hóa tổng quãng đường

Pruning:
  1. Cận dưới = chi phí hiện tại + Σ_j min_in[j] (j chưa thăm)
  2. Khởi tạo upper bound bằng nearest-neighbor
  3. Sắp xếp ứng viên theo chi phí tăng dần
"""

import time
from typing import List, Tuple, Optional


class CBUSBnB:
    def __init__(self, n: int, k: int, cost_matrix: List[List[int]]):
        """
        Khởi tạo solver Branch-and-Bound
        
        Args:
            n: số điểm đón (delivery nodes = 1..n, pickup nodes = n+1..2n)
            k: sức chứa xe
            cost_matrix: ma trận khoảng cách (2n+1) x (2n+1)
        """
        self.n = n
        self.k = k
        self.N = 2 * n + 1  # tổng số nodes (depot + n pickups + n deliveries)
        self.c = cost_matrix
        
        # q[i] = +1 (pickup), -1 (delivery), 0 (depot)
        self.q = [0] * self.N
        for i in range(1, n + 1):
            self.q[i] = 1
        for i in range(n + 1, 2 * n + 1):
            self.q[i] = -1
        
        # Tính min incoming edge cho mỗi node (dùng cho lower bound)
        self.min_in = [float('inf')] * self.N
        for j in range(self.N):
            for i in range(self.N):
                if i != j:
                    self.min_in[j] = min(self.min_in[j], self.c[i][j])
        
        # Kết quả tốt nhất
        self.best_cost = float('inf')
        self.best_route = []
        self.cur_route = []
        
        # Thống kê
        self.start_time = None
        self.time_limit = 60.0
        self.nodes_explored = 0
        self.aborted = False
    
    def time_up(self) -> bool:
        """Kiểm tra hết thời gian"""
        self.nodes_explored += 1
        if (self.nodes_explored & 0xFFFF) != 0:  # check every 65536 nodes
            return self.aborted
        elapsed = time.time() - self.start_time
        if elapsed > self.time_limit:
            self.aborted = True
        return self.aborted
    
    def lower_bound_remaining(self, visited: int) -> int:
        """Tính cận dưới dựa trên các edges chưa thăm"""
        lb = self.min_in[0]
        for j in range(1, self.N):
            if not (visited & (1 << (j - 1))):
                lb += self.min_in[j]
        return lb
    
    def nearest_neighbor_seed(self):
        """Tìm giải pháp khởi tạo bằng nearest-neighbor heuristic"""
        cur = 0
        load = 0
        cost = 0
        visited = 0
        route = []
        
        for step in range(2 * self.n):
            best_j = -1
            best_d = float('inf')
            
            # Tìm node tiếp theo tốt nhất
            for j in range(1, 2 * self.n + 1):
                # Bỏ qua nếu đã thăm
                if visited & (1 << (j - 1)):
                    continue
                
                # Nếu là delivery, kiểm tra pickup đã được thăm chưa
                if j > self.n and not (visited & (1 << (j - self.n - 1))):
                    continue
                
                # Kiểm tra sức chứa
                new_load = load + self.q[j]
                if new_load < 0 or new_load > self.k:
                    continue
                
                # Chọn edge rẻ nhất
                if self.c[cur][j] < best_d:
                    best_d = self.c[cur][j]
                    best_j = j
            
            if best_j == -1:
                return  # Không tìm được giải pháp khả thi
            
            route.append(best_j)
            cost += self.c[cur][best_j]
            visited |= 1 << (best_j - 1)
            load += self.q[best_j]
            cur = best_j
        
        # Quay về depot
        cost += self.c[cur][0]
        
        if cost < self.best_cost:
            self.best_cost = cost
            self.best_route = route[:]
    
    def dfs(self, cur: int, visited: int, load: int, cost: int):
        """
        DFS với pruning
        
        Args:
            cur: node hiện tại
            visited: bitmask các nodes đã thăm
            load: tải trọng xe hiện tại
            cost: chi phí hiện tại
        """
        if self.time_up():
            return
        
        # Pruning: nếu cost >= best_cost thì cắt nhánh
        if cost >= self.best_cost:
            return
        
        # Kiểm tra xem đã thăm hết tất cả nodes
        full = (1 << (2 * self.n)) - 1
        if visited == full:
            total = cost + self.c[cur][0]
            if total < self.best_cost:
                self.best_cost = total
                self.best_route = self.cur_route[:]
            return
        
        # Pruning: nếu lower bound >= best_cost thì cắt nhánh
        if cost + self.lower_bound_remaining(visited) >= self.best_cost:
            return
        
        # Tìm các node ứng viên
        candidates = []
        for j in range(1, 2 * self.n + 1):
            # Bỏ qua nếu đã thăm
            if visited & (1 << (j - 1)):
                continue
            
            # Nếu là delivery, kiểm tra pickup đã được thăm chưa
            if j > self.n and not (visited & (1 << (j - self.n - 1))):
                continue
            
            # Kiểm tra sức chứa
            new_load = load + self.q[j]
            if new_load < 0 or new_load > self.k:
                continue
            
            candidates.append((self.c[cur][j], j))
        
        # Sắp xếp theo chi phí tăng dần (best-first)
        candidates.sort()
        
        # Duyệt DFS
        for _, j in candidates:
            self.cur_route.append(j)
            self.dfs(j, visited | (1 << (j - 1)), load + self.q[j], cost + self.c[cur][j])
            self.cur_route.pop()
            if self.aborted:
                return
    
    def solve(self, time_limit: float = 60.0) -> Tuple[List[int], int, float]:
        """
        Giải bài toán CBUS
        
        Args:
            time_limit: giới hạn thời gian (giây)
        
        Returns:
            (route, best_cost, elapsed_time)
        """
        self.time_limit = time_limit
        self.start_time = time.time()
        self.nodes_explored = 0
        self.aborted = False
        
        # Bước 1: Khởi tạo với nearest-neighbor
        self.nearest_neighbor_seed()
        
        # Bước 2: DFS với pruning
        self.cur_route = []
        self.dfs(0, 0, 0, 0)
        
        elapsed = time.time() - self.start_time
        
        return self.best_route, int(self.best_cost), elapsed


def read_cbus_input(file_path: str) -> Tuple[int, int, List[List[int]]]:
    """Đọc input từ file định dạng: n k + ma trận cost"""
    with open(file_path, 'r') as f:
        lines = f.readlines()
    
    first_line = lines[0].split()
    n = int(first_line[0])
    k = int(first_line[1])
    
    size = 2 * n + 1
    c = []
    idx = 0
    
    for i in range(size):
        if i == 0:
            row = list(map(int, first_line[2:]))
            remaining = size - len(row)
            if remaining > 0:
                row.extend(map(int, lines[1].split()[:remaining]))
            idx = 1
        else:
            start_idx = idx
            while len(c) < i + 1:
                row = list(map(int, lines[start_idx].split()))
                c.append(row)
                start_idx += 1
                if start_idx >= len(lines):
                    break
    
    # Đảm bảo ma trận đúng kích thước
    assert len(c) == size
    for row in c:
        assert len(row) == size
    
    return n, k, c


if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
        time_limit = float(sys.argv[2]) if len(sys.argv) > 2 else 60.0
        
        n, k, c = read_cbus_input(file_path)
        solver = CBUSBnB(n, k, c)
        route, cost, elapsed = solver.solve(time_limit)
        
        print(f"n={n}")
        if route:
            print(" ".join(map(str, route)))
        else:
            print("-1")
        
        # Verbose output
        print(f"# cost={cost}", file=sys.stderr)
        print(f"# time={elapsed:.3f}s", file=sys.stderr)
        print(f"# nodes_explored={solver.nodes_explored}", file=sys.stderr)
        print(f"# status={'TIMEOUT' if solver.aborted else 'OPTIMAL'}", file=sys.stderr)

File: cbus_bnb.cpp
// CBUS - Pickup and Delivery Problem (n < 15)
//
// Bộ giải Branch-and-Bound (DFS + pruning) cho mô hình CBUS.
// Tương đương về kết quả với mô hình MIP trong ToiUuLapKeHoach.pdf
// nhưng không cần thư viện ngoài — chỉ cần g++.
//
// Các ràng buộc được tôn trọng đầy đủ:
//   * Mỗi điểm trong {1..2n} ghé đúng một lần, xuất phát và quay về 0
//   * Pickup i ghé trước Delivery i+n
//   * Sức chứa [0, k] tại mọi bước
//   * Tối thiểu hóa tổng quãng đường
//
// Pruning:
//   1. Cận dưới = chi phí hiện tại + Σ_j min_in[j] (j chưa thăm) + min_in[0]
//      (mỗi điểm sẽ phải được "vào" ít nhất một lần với cạnh rẻ nhất)
//   2. Khởi tạo upper bound bằng nearest-neighbor (có tôn trọng precedence + capacity)
//   3. Sắp xếp ứng viên theo c[cur][j] tăng dần (best-first nội bộ)
//
// I/O:
//   stdin/file:
//     n k
//     ma trận (2n+1) x (2n+1)
//   stdout:
//     n
//     v_1 ... v_{2n}      (-1 nếu không khả thi)
//
// Build:  g++ -O2 -std=c++17 cbus_bnb.cpp -o cbus_bnb
// Run:    ./cbus_bnb [--time SECONDS] [input.txt]

#include <algorithm>
#include <chrono>
#include <climits>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
using Clock = chrono::steady_clock;

struct Solver {
    int n = 0, k = 0, N = 0;
    vector<vector<int>> c;
    vector<int> q;          // +1 pickup, -1 delivery, 0 depot
    vector<int> min_in;     // min incoming edge for each node

    long long best_cost = LLONG_MAX;
    vector<int> best_route;
    vector<int> cur_route;

    Clock::time_point start;
    double time_limit_sec = 60.0;
    bool aborted = false;
    long long nodes_explored = 0;

    bool time_up() {
        if ((nodes_explored & 0xFFFF) != 0) return aborted;
        double e = chrono::duration<double>(Clock::now() - start).count();
        if (e > time_limit_sec) aborted = true;
        return aborted;
    }

    long long lower_bound_remaining(uint32_t visited) const {
        long long lb = min_in[0];
        for (int j = 1; j < N; j++) {
            if (!(visited & (1u << (j - 1)))) lb += min_in[j];
        }
        return lb;
    }

    void dfs(int cur, uint32_t visited, int load, long long cost) {
        ++nodes_explored;
        if (time_up()) return;
        if (cost >= best_cost) return;

        const uint32_t full = (n == 0) ? 0u : ((1u << (2 * n)) - 1u);
        if (visited == full) {
            long long total = cost + c[cur][0];
            if (total < best_cost) {
                best_cost = total;
                best_route = cur_route;
            }
            return;
        }

        if (cost + lower_bound_remaining(visited) >= best_cost) return;

        // gather feasible candidates, sorted by edge cost
        vector<pair<int, int>> cand;
        cand.reserve(2 * n);
        for (int j = 1; j <= 2 * n; j++) {
            if (visited & (1u << (j - 1))) continue;
            if (j > n) {
                int p = j - n;
                if (!(visited & (1u << (p - 1)))) continue;
            }
            int nl = load + q[j];
            if (nl < 0 || nl > k) continue;
            cand.emplace_back(c[cur][j], j);
        }
        sort(cand.begin(), cand.end());

        for (auto [d, j] : cand) {
            cur_route.push_back(j);
            dfs(j, visited | (1u << (j - 1)), load + q[j], cost + d);
            cur_route.pop_back();
            if (aborted) return;
        }
    }

    void nearest_neighbor_seed() {
        int cur = 0, load = 0;
        long long cost = 0;
        uint32_t visited = 0;
        vector<int> route;
        for (int step = 0; step < 2 * n; step++) {
            int bj = -1, bd = INT_MAX;
            for (int j = 1; j <= 2 * n; j++) {
                if (visited & (1u << (j - 1))) continue;
                if (j > n && !(visited & (1u << (j - n - 1)))) continue;
                int nl = load + q[j];
                if (nl < 0 || nl > k) continue;
                if (c[cur][j] < bd) { bd = c[cur][j]; bj = j; }
            }
            if (bj == -1) return;
            route.push_back(bj);
            cost += c[cur][bj];
            visited |= 1u << (bj - 1);
            load += q[bj];
            cur = bj;
        }
        cost += c[cur][0];
        if (cost < best_cost) { best_cost = cost; best_route = route; }
    }

    bool read_input(istream& in) {
        if (!(in >> n >> k)) return false;
        N = 2 * n + 1;
        c.assign(N, vector<int>(N, 0));
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                if (!(in >> c[i][j])) return false;
        q.assign(N, 0);
        for (int i = 1; i <= n; i++) q[i] = 1;
        for (int i = n + 1; i <= 2 * n; i++) q[i] = -1;
        min_in.assign(N, INT_MAX);
        for (int j = 0; j < N; j++) {
            int m = INT_MAX;
            for (int i = 0; i < N; i++) if (i != j) m = min(m, c[i][j]);
            min_in[j] = m;
        }
        return true;
    }

    void solve() {
        start = Clock::now();
        nearest_neighbor_seed();
        cur_route.clear();
        dfs(0, 0u, 0, 0);
    }

    void write_output(ostream& out) const {
        if (best_route.empty()) { out << -1 << "\n"; return; }
        out << n << "\n";
        for (size_t i = 0; i < best_route.size(); i++) {
            if (i) out << " ";
            out << best_route[i];
        }
        out << "\n";
    }
};

int main(int argc, char** argv) {
    Solver s;
    string path;
    bool verbose = false;
    for (int i = 1; i < argc; i++) {
        string a = argv[i];
        if (a == "--time" && i + 1 < argc) s.time_limit_sec = atof(argv[++i]);
        else if (a == "--verbose" || a == "-v") verbose = true;
        else if (!a.empty() && a[0] != '-') path = a;
    }

    if (!path.empty()) {
        ifstream fin(path);
        if (!fin) { cerr << "cannot open " << path << "\n"; return 2; }
        if (!s.read_input(fin)) { cerr << "bad input\n"; return 2; }
    } else {
        if (!s.read_input(cin)) { cerr << "bad input\n"; return 2; }
    }

    s.solve();
    s.write_output(cout);

    if (verbose) {
        double e = chrono::duration<double>(Clock::now() - s.start).count();
        cerr << "# n=" << s.n << " k=" << s.k << "\n";
        cerr << "# objective=" << s.best_cost << "\n";
        cerr << "# wall_time=" << e << "s\n";
        cerr << "# nodes_explored=" << s.nodes_explored << "\n";
        cerr << "# status=" << (s.aborted ? "TIME_LIMIT" : "OPTIMAL") << "\n";
    }
    return 0;
}
